File: OJ/leet721/leet721.cpp
/*
leet721


*/


#include <iostream>
#include <vector>
#include <queue>
#include <stack>
#include <string>
#include <unordered_set>
#include <unordered_map>
#include <algorithm>
#include <tuple>
#include <cassert>
#include <utility>
#include "../utils/LeetCpp.utils.hpp"

using namespace std;




class Solution {
public:
    vector<int> P;
    unordered_map<string, int> m2p;
    unordered_map<int, vector<string>> p2m;
    
    int find(int i) {
        if (P[i] < 0) return i;
        return P[i] = find(P[i]);
    }

    void merge(int i, int j) {
        int ri = find(i), rj = find(j);
        if (ri == rj) return;
        if (P[ri] < P[rj]) P[ri] += P[rj], P[rj] = ri;
        else P[rj] += P[ri], P[ri] = rj;
    }


    vector<vector<string>> accountsMerge(const vector<vector<string>>& accounts) {
        int n = accounts.size();
        P = vector<int>(n, -1);
        for (int i = 0; i < n; ++i) { 
            auto mails = accounts[i];
            for (int j = 1; j < mails.size(); ++j) { 
                if (m2p.count(mails[j]) > 0) merge(m2p[mails[j]], i);
                m2p[mails[j]] = i;
            }
        }

        for (const auto& [mail, i]: m2p) {
            p2m[find(i)].push_back(mail);
        }

        vector<vector<string>> ans;
        for ( auto& [i, mails]: p2m) {
            vector<string> ai = {accounts[i][0]};
            sort(mails.begin(), mails.end());
            ai.insert(ai.end(), mails.begin(), mails.end());
            ans.push_back(ai);
        }
        return ans;
    }
};


int main(int argc, char const *argv[])
{
    Solution sol;   
    cout << sol.accountsMerge({{"David","David0@m.co","David1@m.co"},{"David","David3@m.co","David4@m.co"},{"David","David4@m.co","David5@m.co"},{"David","David2@m.co","David3@m.co"},{"David","David1@m.co","David2@m.co"}}) << endl;
    return 0;
}

File: OJ/leet721/leet721.py
# leet721

from typing import List


class Solution:
    def accountsMerge(self, accounts: List[List[str]]) -> List[List[str]]:
        P = [-1]  * len(accounts)
        def find(i):
            if P[i] < 0: return i 
            P[i] = find(P[i]) 
            return P[i]
        def merge(i, j):
            ri, rj = find(i), find(j)
            if ri == rj: return 
            if P[ri] < P[rj]: P[ri] += P[rj]; P[rj] = ri 
            else: P[rj] += P[ri]; P[ri] = rj 

        mail2pos = {}
        for i, mails in enumerate(accounts):
            mails = mails[1:]
            for mail in mails:
                if mail in mail2pos:
                    merge(i, mail2pos[mail]) 
            for mail in mails:
                mail2pos[mail] = i 
        
        from collections import defaultdict 
        pos2mails = defaultdict(list)
        for mail, i in mail2pos.items():
            pos2mails[find(i)].append(mail)
        
        ans = []
        for key, vals in pos2mails.items():
            ans.append([accounts[key][0], *sorted(vals)])

        return ans



print(Solution().accountsMerge( [["John", "johnsmith@mail.com", "john00@mail.com"], ["John", "johnnybravo@mail.com"], ["John", "johnsmith@mail.com", "john_newyork@mail.com"], ["Mary", "mary@mail.com"]]))

print(Solution().accountsMerge([["David","David0@m.co","David1@m.co"],["David","David3@m.co","David4@m.co"],["David","David4@m.co","David5@m.co"],["David","David2@m.co","David3@m.co"],["David","David1@m.co","David2@m.co"]]))

